src/xdg-autostart-generator/xdg-autostart-service.c

```diff
@@ -0,0 +1,581 @@
+                 * Other implementations may handle this differently.
+                 */
+                if (STR_IN_SET(c,
+                               "%f", "%F",
+                               "%u", "%U",
+                               "%d", "%D",
+                               "%n", "%N",
+                               "%i", /* Location of icon, could be implemented. */
+                               "%c", /* Translated application name, could be implemented. */
+                               "%k", /* Location of desktop file, could be implemented. */
+                               "%v",
+                               "%m"
+                               ))
+                        continue;
+
+                /*
+                 * %% -> % and then % -> %% means that we correctly quote any %
+                 * and also quote any left over (and invalid) % specifier from
+                 * the desktop file.
+                 */
+                raw = strreplace(c, "%%", "%");
+                if (!raw)
+                        return log_oom();
+                p = strreplace(raw, "%", "%%");
+                if (!p)
+                        return log_oom();
+                escaped = cescape(p);
+                if (!escaped)
+                        return log_oom();
+
+                if (asprintf(&quoted, "\"%s\"", escaped) < 0)
+                        return log_oom();
+
+                free(exec_split[n]);
+                exec_split[n++] = TAKE_PTR(quoted);
+        }
+        for (; exec_split[n]; n++)
+                exec_split[n] = mfree(exec_split[n]);
+
+        res = strv_join(exec_split, " ");
+        if (!res)
+                return log_oom();
+
+        *ret_exec_start = res;
+        return 0;
+}
+
+int xdg_autostart_service_generate_unit(
+                XdgAutostartService *service,
+                const char *dest) {
+
+        _cleanup_free_ char *path_escaped = NULL;
+        _cleanup_free_ char *exec_start = NULL;
+        _cleanup_fclose_ FILE *f = NULL;
+        const char *unit;
+        int r;
+
+        assert(service);
+
+        /* Nothing to do for hidden services. */
+        if (service->hidden) {
+                log_info("Not generating service for XDG autostart %s, it is hidden", service->name);
+                return 0;
+        }
+
+        if (service->systemd_skip) {
+                log_info("Not generating service for XDG autostart %s, should be skipped by generator", service->name);
+                return 0;
+        }
+
+        /* Nothing to do if type is not Application. */
+        if (!streq_ptr(service->type, "Application")) {
+                log_info("Not generating service for XDG autostart %s, it is hidden", service->name);
+                return 0;
+        }
+
+        if (!service->exec_string) {
+                log_info("Not generating service for XDG autostart %s, it is has no Exec= line", service->name);
```

Comment:
this one is probably more than an informational message no? i.e. a desktop file must have an Exec= line to be valid, no? If so, maybe upgrade this line to log_warning()?